### PR TITLE
Constraint code homogenization

### PR DESCRIPTION
**Pull request by rsanchezsaez**
I went over all classes with constraints and homogenized the constraint code (this is a big pull request). This partially deals with [Issue #110](https://github.com/ResearchKit/ResearchKit/issues/110), and is related to the comments in [Issue #99](https://github.com/ResearchKit/ResearchKit/issues/99).

While I was doing this, I tested every individual class immediately after changing it. I tried not to break anything, but a full UI review would be a good idea.

It takes care of the following items (some are purely stylistic, but others slightly affect how the code works):

*Stylistic changes*
- Always use the property format when doing `view.translatesAutoresizingMaskIntoConstraints = NO` instead of the method format (`setTranslatesAutoresizingMaskIntoConstraints:`).
- Align all constraint code so it doesn't overflow the screen (easier to read).
- Use `CGFloat` literals (`1.0` and `0.0`) on `multiplier` and `constant` parameters when creating constraints. This visually indicates that these are `CGFloat`.
- Use the `ORKScreenMetricMaxDimension` constant instead of magic numbers for constraints requesting a very large width or height.
- Use `(NSLayoutFormatOptions)0` instead of just `0` when passing constraint options (makes argument clearer).
- Use `[array count]` method syntax instead of `array.count` property syntax.

*Code changes*
- All constraints are added and removed by activation/deactivation (`NSLayoutConstraint activateConstraints:` and `NSLayoutConstraint deactivateConstraints`). This is the [preferred way of doing it on *iOS 8* and up](https://developer.apple.com/library/ios/documentation/AppKit/Reference/NSLayoutConstraint_Class/#//apple_ref/occ/instp/NSLayoutConstraint/active) as it adds the constraints at the most optimal place.
- Make sure static constraints are created only once, usually in the `-setUpConstraints` method.
- Variable constraints (which need to be recalculated) are always stored in the `_variableConstraints` ivar. These are always removed and recreated in the same way.
- Constraints which need *constant* updating are stored as ivars (instead of in a `NSDictionary`): less boilerplatery and more type-safe.
- Constraints which need *constant* updating are updated on the `updateConstraintsConstantForWindow:` method. This method is usually called: when there are changes in the view; in `-updateConstraints`; and in `-willMoveToWindow:`. The last one is the reason of the window argument.
- Remove all `_screenType` ivars, which was needless artificial state (error prone). Instead, the constraints are updated correctly by passing the appropriate `UIWindow` to `updateConstraintsConstantForWindow:`.
- `ORKImageSelectionView` structure was simplified by dropping the invisible label (the choice label takes the functionality of making sure the `intrinsicContentSize` is tall enough for the tallest string choice).
- Use internal *ivars* instead of internal properties in some classes (for coherence with most of the code).

I may be forgetting other things that I did, I'll update the PR if something comes up. Let me know if you have any question.

Although these changes will be quite a task to review, I think they make constraint code more homogenous, more stable, easier to read, and easier to maintain.

### CODE REVIEW

**Comment by rsanchezsaez** [comment]
I couldn't help myself and, taking advantage of this big PR, I refactored the uses of `array.count` property syntax into the more explicit `[array count]` method syntax. I'll add this recommendation to the *Coding Style Guide* as well.

**Comment by rsanchezsaez** [comment]
Hmm, regarding the last commit on this PR: I just realized that according to the documentation `count` changed from a method to a property on *iOS 8* (I guess this was done for coherence with Swift). Should we then favour the `.count` syntax instead on *Objective-C*?

I can easily overwrite the last commit with the opposite style, and update the *Coding Style Guide* accordingly.

**Comment by rsanchezsaez** [comment]
I deleted the `[array count]` commit from this PR to facilitate reviewing. I'll submit it as a different PR, but with the (now officially documented) `array.count` format. (Unless you tell me otherwise).

**Comment by YuanZhu-apple** [comment]
@rsanchezsaez I think we should follow `.count` syntax in this PR.

**Comment by rsanchezsaez** [comment]
Ok! I'll add some property syntax refactoring commits (for `count`, `length`, `*Value`, and a few others) to this PR.

**Comment by rsanchezsaez** [comment]
Added.

**Comment by YuanZhu-apple** [comment]
@rsanchezsaez Thanks you the code refactor!
Do you think by any chance you can address  #290 "Implement size class support" in this PR?

**Comment by rsanchezsaez** [comment]
Yeah, I'll look into that in a few days.

**Comment by YuanZhu-apple** [comment]
@rsanchezsaez I know you are also taking on the **Charts** PR, do you think you can fix #290 by early next week? 

Or I can create a separate PR to address it. Then this one can be merge with my PR.

**Comment by rsanchezsaez** [comment]
It's a bit of a tight schedule, I'm not sure if I'll have time to do it so quick.

Could you possibly give it a stab starting from my *rsanchezsaez-constraints* branch? That way the merge would be super-easy.

---

Note that the `deactivateConstraints:` *iOS 9* bug discussed [in the ORKTowerOfHanoiView crash PR](https://github.com/ResearchKit/ResearchKit/pull/362) will probably affect other parts of this refactoring. Let's hope they fix it at the *iOS* level soon.

**Comment by YuanZhu-apple** [comment]
I like to make my PR base on your branch, but I don't think it would compile under Xcode7(ios9)?
The main purpose of my PR is to make RK looks nicer under multitask split view in iOS9.

**Comment by rsanchezsaez** [comment]
Good point. Then you could make it against the *iOS 9* branch, and I'll figure how to merge it with my constraint refactoring branch later.

**Comment by rsanchezsaez** [comment]
Hey @YuanZhu-apple, do you think is it a good idea to try and merge this early in the release cycle? That way we'll have time to catch and fix any regressions, if there's any.

**Comment by YuanZhu-apple** [comment]
Sure, I will review this.

File: ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m [hunk 1]
```diff
@@ -111,91 +110,89 @@ - (void)setImage:(UIImage *)image {
...
-    NSLayoutConstraint *zeroWidthConstraint = [NSLayoutConstraint constraintWithItem:self
-                                                                           attribute:NSLayoutAttributeWidth
-                                                                           relatedBy:NSLayoutRelationEqual
-                                                                              toItem:nil
-                                                                           attribute:NSLayoutAttributeNotAnAttribute
-                                                                          multiplier:1.0
-                                                                            constant:0.0];
-    zeroWidthConstraint.priority = UILayoutPriorityRequired-1;
-    _zeroWidthConstraint = zeroWidthConstraint;
-    _zeroWidthConstraint.active = !_enabled;
```
Comment (YuanZhu-apple) [inline ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m]: We don't need this line now?

File: ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m [hunk 2]
```diff
@@ -245,12 +242,12 @@ - (instancetype)initWithFrame:(CGRect)frame {
         [self addSubview:_leftView];
         [self addSubview:_rightView];
         [self addSubview:_metricKeyline];
-        [self setupConstraints];
+        [self setUpConstraints];
```
Comment (YuanZhu-apple) [inline ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m]: Will this code make a loop?

`setUpConstraints` -> `setNeedsUpdateConstraints` -> `updateConstraints` -> `setUpConstraints`

File: ResearchKit/ActiveTasks/ORKDataLogger.m [hunk 3]
```diff
@@ -1075,13 +1075,17 @@ - (instancetype)initWithDirectory:(NSURL *)directory delegate:(id<ORKDataLoggerM
     return self;
 }
 
+ORKDefineStringKey(PendingUploadBytesThresholdKey);
+ORKDefineStringKey(TotalBytesThresholdKey);
+ORKDefineStringKey(LoggerConfigurationsKey);
```
Comment (YuanZhu-apple) [inline ResearchKit/ActiveTasks/ORKDataLogger.m]: Would this renaming cause a migration on the existing archive?

File: ResearchKit/ActiveTasks/ORKSpatialSpanMemoryContentView.m [hunk 4]
```diff
@@ -274,29 +272,25 @@ - (void)setBounds:(CGRect)bounds {
...
+                                             metrics:nil
+                                               views:views]];
+    NSLayoutConstraint *gameViewHeightConstraint = [NSLayoutConstraint constraintWithItem:_gameView
+                                                                      attribute:NSLayoutAttributeHeight
+                                                                      relatedBy:NSLayoutRelationEqual
+                                                                         toItem:nil
+                                                                      attribute:NSLayoutAttributeNotAnAttribute
+                                                                     multiplier:1.0
+                                                                       constant:ORKScreenMetricMaxDimension];
+    gameViewHeightConstraint.priority = UILayoutPriorityDefaultLow-1;
```
Comment (YuanZhu-apple) [inline ResearchKit/ActiveTasks/ORKSpatialSpanMemoryContentView.m]: Not your code, add spaces around `-`

File: ResearchKit/Common/ORKImageCaptureStepViewController.m [hunk 5]
```diff
@@ -109,26 +121,26 @@ - (void)retakePressed:(void (^ __nullable)())handler {
...
 }
 
 - (void)videoOrientationDidChange:(AVCaptureVideoOrientation)videoOrientation {
     // Keep the output orientation in sync with the input orientation
-    ((AVCaptureConnection *)self.stillImageOutput.connections[0]).videoOrientation = videoOrientation;
+    ((AVCaptureConnection *)_stillImageOutput.connections[0]).videoOrientation = videoOrientation;
 }
 
-- (void)queue_captureImageFromData:(CMSampleBufferRef)imageDataSampleBuffer withHandler:(void (^ __nullable)(BOOL))handler {
+- (void)queueCaptureImageFromData:(CMSampleBufferRef)imageDataSampleBuffer handler:(void (^)(BOOL))handler {
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKImageCaptureStepViewController.m]: Can you leave the `_` in the middle, I find it is easy to read.

File: ResearchKit/Common/ORKSurveyAnswerCell.m [hunk 6]
```diff
@@ -143,11 +143,11 @@ - (void)keyboardWillAppear:(NSNotification *)aNotification {
     _cachedContentInsets = tableView.contentInset;
     _cachedScrollIndicatorInsets = tableView.scrollIndicatorInsets;
     
-    NSDictionary *info = [aNotification userInfo];
-    CGSize kbSize = [info[UIKeyboardFrameEndUserInfoKey] CGRectValue].size;
-    kbSize.height = kbSize.height-44;
+    NSDictionary *userInfo = aNotification.userInfo;
+    CGSize keyboardSize = ((NSValue *)userInfo[UIKeyboardFrameEndUserInfoKey]).CGRectValue.size;
+    keyboardSize.height = keyboardSize.height-44;
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKSurveyAnswerCell.m]: Space around `-`

File: ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m (hunk 2 above)
Comment (rsanchezsaez) [inline ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m]: This `setUpConstraints` call is on the `init` method, and `updateConstraints` doesn't call `setUpConstraints`. I don't think there's an infinite loop, could you re-check?

File: ResearchKit/Common/ORKImageCaptureStepViewController.m (hunk 5 above)
Comment (rsanchezsaez) [inline ResearchKit/Common/ORKImageCaptureStepViewController.m]: Are you sure? Isn't this very much against the [Coding Guidelines for Cocoa](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html#//apple_ref/doc/uid/20001282-BCIGIJJF)?

File: ResearchKit/ActiveTasks/ORKDataLogger.m (hunk 3 above)
Comment (rsanchezsaez) [inline ResearchKit/ActiveTasks/ORKDataLogger.m]: Good catch, I'm now reusing the old keys for backwards compatibility.

**Comment by rsanchezsaez** [comment]
I have taken care of some of the comment, replied to others, and added some extra *Coding Style Guide* conformance commits.

File: ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m (hunk 1 above)
Comment (rsanchezsaez) [inline ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m]: Activation of the `_zeroWidthConstraint` constraint is taken care in `updateConstraints`, which will take place at the most optimal time after `setNeedsUpdateConstraints` is called.

**Comment by rsanchezsaez** [comment]
I have added a bunch of *Coding Style Guide* homogenization commits.

File: ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m (hunk 2 above)
Comment (YuanZhu-apple) [inline ResearchKit/ActiveTasks/ORKActiveStepQuantityView.m]: Nevermind, my bad, I thought this is from `updateConstraints`.

File: ResearchKit/Common/ORKImageCaptureStepViewController.m (hunk 5 above)
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKImageCaptureStepViewController.m]: I just think this falls into a special use case, to me and @umerkhan-apple,  `_` is more visible.

Comment (rsanchezsaez) [inline ResearchKit/Common/ORKImageCaptureStepViewController.m]: Sure thing, done. It makes kind of sense for methods that are only called from a *GCD* queue.

**Comment by YuanZhu-apple** [comment]
I found a issue which might not from this PR.


1. Launch ORKTest app in iPad simulator in landscape mode
2. Rotate simulator to Portrait mode
3. All the items form into one column, which is unexpected.

screen shot 2015-09-22 at 5 08 07 pm

**Comment by rsanchezsaez** [comment]
Yeah, that's an unrelated problem. I have submitted a separate [PR](https://github.com/ResearchKit/ResearchKit/issues/474) with a fix.

**Comment by YuanZhu-apple** [comment]
r=me. Thank you @rsanchezsaez !

**Comment by rsanchezsaez** [comment]
Excellent! Let me know if any UI issue comes up in the next full UI review.

**Pull request merged**